### PR TITLE
INFRA-1677 Speedup parsing file

### PR DESCRIPTION
**Pull request by cubasepp**
before:

```
# dev@mnb-00142: ~/work/metoda-infra/apps/orchparty <main ✘ [*?]>                                                                                                            shared 
~ ORCH_ENV=dev bin/console                                                                                                                                                 10:51:31
irb(main):001:0> file = '/home/michael/work/metoda-infra/stacks/metoda-account/stack.rb'
=> "/home/michael/work/metoda-infra/stacks/metoda-account/stack.rb"
irb(main):002:0> Orchparty.upgrade(cluster_name: 'test', application_name: 'metoda-account', force_variable_definition: nil, file_name: file)
       user     system      total        real
   0.000405   0.000031   0.000436 (  0.000434)
   0.087585   0.000000   0.087585 (  0.087682)
   7.404064   0.000000   7.404064 (  7.408074)
```

after:

```
=> "/home/michael/work/metoda-infra/stacks/metoda-account/stack.rb"
irb(main):002:0> Orchparty.upgrade(cluster_name: 'kind-metoda-infra', application_name: 'metoda-account', force_variable_definition: nil, file_name: file)
       user     system      total        real
   0.000381   0.000041   0.000422 (  0.000418)
   0.090710   0.000000   0.090710 (  0.090858)
   0.383288   0.000000   0.383288 (  0.383700)
```

### CODE REVIEW

File: lib/orchparty/transformations/variable.rb [hunk 1]
```diff
@@ -8,24 +8,27 @@ def initialize(opts = {})
 
       def transform(ast)
         ast.applications.each do |_, application|
+          ctx = build_context(application: application)
```
Comment (cubasepp) [inline lib/orchparty/transformations/variable.rb:11]: Build the default context.

File: lib/orchparty/transformations/variable.rb [hunk 2]
```diff
@@ -8,24 +8,27 @@ def initialize(opts = {})
...
           application.volumes = application.volumes.each do |_, volume|
-            resolve(application, volume, nil) if volume
+            resolve(ctx, volume) if volume
           end
         end
         ast
       end
 
-      def resolve(application, subject, service)
+      def resolve(ctx, subject)
```
Comment (cubasepp) [inline lib/orchparty/transformations/variable.rb:22]: reuse context

File: lib/orchparty/transformations/variable.rb [hunk 3]
```diff
@@ -8,24 +8,27 @@ def initialize(opts = {})
...
         subject.deep_transform_values! do |v|
           if v.respond_to?(:call)
-            eval_value(build_context(application: application, service: service), v) 
+            ctx = merge_service(ctx, subject)
```
Comment (cubasepp) [inline lib/orchparty/transformations/variable.rb:25]: only merge the current service/subject not the whole file... is faster.

File: lib/orchparty/transformations/variable.rb [hunk 4]
```diff
@@ -40,13 +43,17 @@ def eval_value(context, value)
         context.instance_exec(&value)
       end
 
-      def build_context(application:, service:)
+      def merge_service(ctx, service)
+        return ctx if service._variables.nil?
```
Comment (cubasepp) [inline lib/orchparty/transformations/variable.rb:47]: needed for volumes.

File: spec/input/variable_example.rb [hunk 5]
```diff
@@ -1,5 +1,6 @@
+require 'byebug'
```
Comment (timonugmann) [inline spec/input/variable_example.rb]: Debugging leftover?

File: spec/input/variable_example.rb [hunk 6]
```diff
@@ -9,31 +10,47 @@
...
+      var lb_config: -> { {
+        "port": web_port,
+        "tags": ["star_metoda"],
+        "fqdns": %w[domain.com]
+      } }
     end
+    lb_config -> { lb_config }
   end
 
 end
```
Comment (timonugmann) [inline spec/input/variable_example.rb]: :)

File: spec/transformations/variables_spec.rb [hunk 7]
```diff
@@ -1,17 +1,18 @@
 require "spec_helper"
-
+require 'byebug'
```
Comment (timonugmann) [inline spec/transformations/variables_spec.rb]: Also here

**Review by timonugmann — APPROVED** [review]
Overall looks good!

**Comment by florin555** [comment]
Nice! How were you able to identify the the source of the slowness?

**Comment by cubasepp** [comment]
@florin555 with the benchmark.

Then later I went deeper into the functions with benchmarks, later I figured out is the merge hell.

Used rspec a lot and rails console.

At the end it was luck to find this.

**Pull request merged**